### PR TITLE
[redgifs] Add extractor

### PR DESCRIPTION
**Pull request by chio0hai**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Add extractor for redgifs.com. It is based on the gfycat extractor.
The generic extractor was not reliable for redgifs.com because:
- It downloads all the videos of different resolution instead of just downloading the best one.
- It does not work on some links as reported by https://github.com/yt-dlp/yt-dlp/issues/1504

### CODE REVIEW

File: yt_dlp/extractor/redgifs.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+
+from ..utils import (
+    ExtractorError,
+    int_or_none,
+    qualities,
+)
+
```
Comment (pukkandan) [inline yt_dlp/extractor/redgifs.py:12]:
```suggestion

from .common import InfoExtractor
from ..utils import (
    ExtractorError,
    int_or_none,
    qualities,
)

```

File: yt_dlp/extractor/redgifs.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url).lower()
+
+        video_info = self._download_json(
+            'https://api.redgifs.com/v2/gifs/%s' % video_id,
+            video_id, 'Downloading video info')
+        if 'error' in video_info:
+            raise ExtractorError('RedGifs said: ' + video_info['error'], expected=True)
```
Comment (pukkandan) [inline yt_dlp/extractor/redgifs.py:55]: Dont use +

```suggestion
            raise ExtractorError(f'RedGifs said: {video_info["error"]}, expected=True)
```
or
```py
            raise ExtractorError('RedGifs said: %s' % video_info['error'], expected=True)
```

File: yt_dlp/extractor/redgifs.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+        gif = video_info['gif']
+        urls = gif['urls']
+
+        # redgifs do not have title
+        title = ' '.join(gif.get('tags')) or 'RedGifs'
```
Comment (pukkandan) [inline yt_dlp/extractor/redgifs.py:61]:
```suggestion
        title = ' '.join(gif.get('tags') or []) or 'RedGifs'
```

File: yt_dlp/extractor/redgifs.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        view_count = int_or_none(gif.get('views'))
+        like_count = int_or_none(gif.get('likes'))
+        age_limit = 18
+
+        orig_width = int_or_none(gif.get('width'))
+        orig_height = int_or_none(gif.get('height'))
+
+        duration = int_or_none(gif.get('duration'))
+
+        categories = gif.get('tags') or []
```
Comment (pukkandan) [inline yt_dlp/extractor/redgifs.py:73]: put all vars that are actually never used directly in the return statement

File: yt_dlp/extractor/redgifs.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+        FORMATS = ('gif', 'sd', 'hd')
```
Comment (pukkandan) [inline yt_dlp/extractor/redgifs.py:75]: makes sense as a class var - like this:

```
formats = {
    'gif': 250,
    'sd': 480,
    'hd': None,
}

File: yt_dlp/extractor/redgifs.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            width = orig_width if height == orig_height else int(height / orig_height * orig_width)
+
+            formats.append({
+                'url': video_url,
+                'format_id': format_id,
+                'width': width,
+                'height': height,
+                'quality': quality(format_id),
+            })
+        self._sort_formats(formats)
```
Comment (pukkandan) [inline yt_dlp/extractor/redgifs.py:100]:
```suggestion
        aspect_ratio = try_get(gif, lambda x: orig_height/x['width'])
        formats = []
        for format_id, height in self._FORMATS.items():
            video_url = urls.get(format_id)
            if not video_url:
                continue
            height = min(orig_height, height or orig_height)
            formats.append({
                'url': video_url,
                'format_id': format_id,
                'width': height * aspect_ratio if aspect_ratio else None,
                'height': height,
                'quality': quality(format_id),
            })
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by u-spec-png** [comment]
related #1504

**Comment by chio0hai** [comment]
The requested changes have been applied.

File: yt_dlp/extractor/redgifs.py (hunk 1 above)
Comment (chio0hai) [reply yt_dlp/extractor/redgifs.py]: Done.

File: yt_dlp/extractor/redgifs.py (hunk 2 above)
Comment (chio0hai) [reply yt_dlp/extractor/redgifs.py]: Done.

File: yt_dlp/extractor/redgifs.py (hunk 3 above)
Comment (chio0hai) [reply yt_dlp/extractor/redgifs.py]: Done.

File: yt_dlp/extractor/redgifs.py (hunk 4 above)
Comment (chio0hai) [reply yt_dlp/extractor/redgifs.py]: Done.

File: yt_dlp/extractor/redgifs.py (hunk 5 above)
Comment (chio0hai) [reply yt_dlp/extractor/redgifs.py]: Done.

File: yt_dlp/extractor/redgifs.py (hunk 6 above)
Comment (chio0hai) [reply yt_dlp/extractor/redgifs.py]: Done.

**Comment by Deer-Spangle** [comment]
Just a head's up, I am eagerly awaiting this extractor, and once it is in, I'll update my PR: #1808 which adds another extractor for RedGifs searches as playlists

**Pull request merged**